[Tags Feed] Implement Post Item UI improvements

thomashorta opened the pull request:
Fixes #20788 

Implement UI improvements to the post item in Tags Feed, making it more compact.

-----

## To Test:

1. Open Jetpack
2. Make sure the `reader_tags_feed` Feature Config is ON
3. Go to Reader
4. Select `Tags` feed
5. **Verify** the post items have the new design

-----

## Regression Notes

1. Potential unintended areas of impact

    - N/A

6. What I did to test those areas of impact (or what existing automated tests I relied on)

    - N/A

7. What automated tests I added (or what prevented me from doing so)

    - N/A, UI changes only

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist (strike-out the not-applying and unnecessary ones):

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [x] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

Commit:
Update Loaded item UI

Commit:
Update height of the "more from" button

Commit:
Update post loading skeleton

Commit:
Improve excerpt rendering and item height

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

develric commented on WordPress/src/main/java/org/wordpress/android/ui/reader/views/compose/tagsfeed/ReaderTagsFeedPostListItem.kt:
```diff
@@ -29,20 +31,27 @@ import androidx.compose.runtime.remember
 import org.wordpress.android.ui.compose.unit.Margin
+import kotlin.math.min
+
+private const val CONTENT_TOTAL_LINES_WITH_INTERACTIONS = 3
+private const val CONTENT_TOTAL_LINES_NO_INTERACTIONS = 4
```
Sorry if it was discussed before 🙇 ...If I get it right, when there are interactions, and the titel is 2 lines, only 1 line remains for excerpt. Should we consider increasing the values below like this?

```kotlin
private const val CONTENT_TOTAL_LINES_WITH_INTERACTIONS = 4
private const val CONTENT_TOTAL_LINES_NO_INTERACTIONS = 5
``` 

IINW that would bring excerpt allowed lines to 2 (with interactions and 2 lines title) and 3 (when no interactions).

I understand it's not easy to fit all the scenarios and from another point of view I appreciate the cards being lighter with less text, what I'm thinking is the cases where we have 1 line for the excerpt might be a bit short context but don't have a real strong opinion; wdyt? (also pinging @osullivanchris for thoughts in case 🙇 )

thomashorta replied:
Good point @develric! I just tested it with these values and it looks more balanced though I needed to do some tweaking in some spacing and positioning to make sure all cases look good.

The main change I needed to do was:
- Title + excerpt now try to fill all the height available
- Excerpt is centered vertically **when it has the maximum number of lines expected**
- Excerpt is aligned to the top, with an `8dp` top margin **when it has fewer lines than the expected number of lines**

This also makes sure the feature image position is more consistent. Before, the image was aligned to the title + excerpt component, so if it was small, the image would be higher up on the card.

Here are some examples:
Title with 2 lines, no likes/comments:
![image](https://github.com/wordpress-mobile/WordPress-Android/assets/5091503/b51d0934-6ef1-4e68-b0fc-ebec5c6ed2e1)

Title with 2 lines, with likes/comments:
![image](https://github.com/wordpress-mobile/WordPress-Android/assets/5091503/b08dbce3-65f4-4d28-9183-305e32b6c624)

Title with 2 lines, no likes/comments, with short excerpt:
![image](https://github.com/wordpress-mobile/WordPress-Android/assets/5091503/8af5f69b-9f56-40a8-8d96-c65e655ccb68)


Title with 1 line, no likes/comments:
![image](https://github.com/wordpress-mobile/WordPress-Android/assets/5091503/e8c3e747-e5ed-4970-99d9-0d44f62b14cc)

Title with 1 line, with likes/comments:
![image](https://github.com/wordpress-mobile/WordPress-Android/assets/5091503/010bfa56-f36d-476b-89e2-2799bbc9e4ec)

Title with 1 line, with likes/comments, with short excerpt:
![image](https://github.com/wordpress-mobile/WordPress-Android/assets/5091503/8174ea13-f6b0-488f-a41f-71a7464a300f)


---

Another thing I noticed is that since we are always fetching the most recent posts for each tag, **most posts don't have likes and **comments**. And since refreshing the page will almost always (depending on the tag of course) bring new posts, this is always true.

Note that I have 7 followed tags which are pretty generic, so they have posts published with them very frequently, and it's actually hard finding posts with likes/comments.

That said, @osullivanchris is the like/comment information useful in this feed? I feel it is making the spacing/position logic for the other elements a bit more complex while being hidden most of the time.

osullivanchris replied:
@thomashorta thanks for all those thoughts. Taking the discussion about number of lines of title and excerpt to Slack to quickly resolve. I actually want it to be more restrictive, taking up a pretty defined area lining up with the image. Couple others:

- in my design I also thought that we might have too many items between username, title, excerpt, metadata, actions. And I had asked if anyone thought there was one we could drop. Given you've made a case to drop the metadata, I am open to it for sure.
- You did give me a feature idea. Obviously separate to this effort. But what if in the default tags feed those were most liked posts or posts with high likes ratio from the last 1-7 days (depending how often/recent the user visits) and then the full chronological stream is visible when the user taps to filter by tag, or to see all the posts from that tag? It might give a better taste of good content here.

thomashorta replied:
> But what if in the default tags feed those were most liked posts or posts with high likes ratio from the last 1-7 days (depending how often/recent the user visits) and then the full chronological stream is visible when the user taps to filter by tag, or to see all the posts from that tag? It might give a better taste of good content here.

That would be interesting. We just need to make sure the content doesn't get stale, so refreshing the screen/coming to it every day actually loads new content.

Regarding the text lines question, as per our discussion on Slack (p1715708170826399/1715699905.724819-slack-C05N140C8H5), I will go with this for now:
- 3 lines for title + excerpt (2 lines for title max)
- 64x64 image
- Vertically centered “content” block
- Keeping the like/comments metadata
    - I’m doing this because it’s easier to keep it for now and then decide to remove it than remove it now and decide to add it again

cc @develric my latest push already implements these changes.

develric reviewed (commented):
Overall code looks good and works well (and btw, thanks for extracting UI constants management into a dedicated utils 🙇).

For visibility (and double-check), AFAIU the requirements in the [linked internal slack](https://github.com/wordpress-mobile/WordPress-Android/issues/20788#issue-2287828281) are:
- make cards width 80% of screen width (✅)
- trying to have cards at constant hight (not straightforward for posts having/not-having images or even title and such) by distributing a fixed hight between title and excerpt (✅)

I think that works as expected 👍 ! I just left a comment [here](https://github.com/wordpress-mobile/WordPress-Android/pull/20802#discussion_r1598134784) to double check with you and Chris if that makes sense, but after that I'm fine with approving/merging this one 👍

Commit:
Improve number of title + excerpt lines

Commit:
Add blog name fallback in case of blank names

Commit:
Change tags feed item content total lines to 3

At most 2 for title, an the rest for excerpt, vertically aligning the content
block between the site tile row and bottom metadata/buttons

Commit:
Update loading skeleton to match new layout

Commit:
Merge branch 'feature/tags-ia' of github.com:wordpress-mobile/WordPress-Android into issue/20788-tags-ia-ui-improvements

Commit:
Bump thin line in skeleton to 10dp

Commit:
Fix unit tests

develric reviewed (approved):
Thanks for the extra thoughts and all the nice tweaks @thomashorta  and @osullivanchris  🙇 ! It looks nice! 

I also like how the skeleton now blends better into the actual content. I still have small doubt about when we show one line of excpert but overall it looks pretty balanced in all the various scenarios; moreover atm I do not have a better suggestion respect the one line excpert scenario and how to balance it when we also have images to align and such so I'm happy to approve the PR, and we can iterate on it later by looping in any user feedback we might get 👍  !

osullivanchris reviewed (approved):
Looks good!!
